### PR TITLE
F/game over

### LINKED ISSUE #41: Game over mechanism
* go game over the minute an NPC collides

### PR DESCRIPTION
**Pull request by tedgag**
Closes #56 
- Added NPC punch mechanic when they are close to the player. The NPC will continuously try to punch the player while chasing when they are close enough to them
- If they succeed, the player loses 1 HP (out of 3), sobers up and is pushed away from the NPC. They are also invincible for a short amount of time, indicated by the flickering effect on the player. The NPC goes back to wandering after since they got their revenge, and can't be aggro'd for a short time
- The player can dodge the attacks of the NPC if they don't collide with their fist when they do the punch animation
- If the player loses all their HP, they get knocked out and the game is over. A short animation plays and the player is sent to the game over scene.
- Current score is saved upon game over

### CODE REVIEW

File: Assets/Scripts/Drinkables/AbstractDrinkable.cs [hunk 1]
```diff
@@ -19,6 +19,7 @@ public abstract class AbstractDrinkable : MonoBehaviour
         protected void CommonDrunkennessEffects()
         {
             // All beers will have a base score multiplier of 2
+            // TODO: Update to base score multiplier directly off intoxication level instead.
```
Comment (ClaudiaLapalme) [inline Assets/Scripts/Drinkables/AbstractDrinkable.cs]: Why would you move this to the intoxication level? Doesn't the multiplier only change when the player drinks?

File: Assets/Scripts/NPCs/NPCManager.cs [hunk 2]
```diff
@@ -55,4 +90,63 @@ public float GetDistanceWithPlayer()
...
+     * If the player collides with a NPC that is not chasing him, start chase.
+     * For now, the collider is of type trigger because there is no navmesh.
+     */
+    void OnCollisionEnter(Collision collision)
+    {
+        if (collision.gameObject.CompareTag("Player"))
+        {
+            if (_canChase && !IsChasing())
+            {
+                if (bumpingSounds.Length > 0)
```
Comment (ClaudiaLapalme) [inline Assets/Scripts/NPCs/NPCManager.cs]: Can you reduced the amount of nesting please

File: Assets/Scripts/NPCs/NPCManager.cs [hunk 3]
```diff
@@ -55,4 +90,63 @@ public float GetDistanceWithPlayer()
...
+        }
+    }
+
+    public void StopPunching()
+    {
+        Punching = false;
+    }
+    public void OnPlayerHit()
+    {
+        _stateMachine.StartWandering();
```
Comment (ClaudiaLapalme) [inline Assets/Scripts/NPCs/NPCManager.cs:114]: After punching the player, I don't think the NPCs should start wandering since their goal is to catch the player. They should maybe have a small cooldown and restart chasing after.

File: Assets/Scripts/Player/PlayerController.cs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            _audioSource = GetComponent<AudioSource>();
+            _animator = GetComponent<Animator>();
+            
+            Indestructibles.Volume.profile.TryGetSettings(out _vignette);
+        }
+
+        private void Update()
+        {
+            // drink a regular beer every time the player presses U
+            if (Input.GetKeyDown(KeyCode.U))
```
Comment (ClaudiaLapalme) [inline Assets/Scripts/Player/PlayerController.cs]: We should probably pair this with a debug bool

File: Assets/Scripts/Player/PlayerController.cs [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+                    if (Indestructibles.PlayerData.IntoxicationLevel < 0.0f)
+                        Indestructibles.PlayerData.IntoxicationLevel = 0.0f;
+                    _animator.SetFloat(IntoxicationHash, Indestructibles.PlayerData.IntoxicationLevel);
+                    if (_vignette != null)
+                    {
+                        _vignette.intensity.value = Indestructibles.PlayerData.IntoxicationLevel;
+                    }
+                }
+                
+            }
```
Comment (ClaudiaLapalme) [inline Assets/Scripts/Player/PlayerController.cs]: This code also seems to be in the `NPCManager` class. Will this not cause it to occur twice simultaneously?

**Review by ClaudiaLapalme — CHANGES_REQUESTED** [review]
I didn't play the code yet, but here's a small review in the meantime

File: Assets/Scripts/Player/PlayerController.cs (hunk 5 above)
Comment (tedgag) [reply Assets/Scripts/Player/PlayerController.cs]: I don't think it is, the OnCollisionEnter in NPCManager is for when the player bumps into the NPCs when they are wandering so they can start the chase, there are not conflicts

File: Assets/Scripts/NPCs/NPCManager.cs (hunk 2 above)
Comment (tedgag) [reply Assets/Scripts/NPCs/NPCManager.cs]: Done

File: Assets/Scripts/Drinkables/AbstractDrinkable.cs [hunk 6]
```diff
@@ -20,8 +20,7 @@ protected void CommonDrunkennessEffects()
         {
             // All beers will have a base score multiplier of 2
             Indestructibles.PlayerData.ScoreMultiplier *= 2;
-            if (Indestructibles.PlayerData.IntoxicationLevel < 1.0f)
```
Comment (ClaudiaLapalme) [inline Assets/Scripts/Drinkables/AbstractDrinkable.cs:23]: why did you remove the if? the camera settings can't go above 1 and this is the workaround to prevent the camera from being completely sober when the player still has beers in their system

File: Assets/Scripts/Indestructibles.cs [hunk 7]
```diff
@@ -2,23 +2,41 @@
 using Structs;
 using UnityEngine;
 using UnityEngine.Rendering.PostProcessing;
+using UnityEngine.SceneManagement;
 
 // General variables that we only need one instance of and that are present in multiple files
 public static class Indestructibles
 {
     // Camera stuff
-    public static readonly PostProcessVolume Volume = GameObject.Find("Camera").GetComponent<PostProcessVolume>();
```
Comment (ClaudiaLapalme) [inline Assets/Scripts/Indestructibles.cs:10]: Why were these variables changed from `readonly` ?

File: Assets/Scripts/NPCs/NPCManager.cs [hunk 8]
```diff
@@ -7,29 +10,61 @@ public class NPCManager : MonoBehaviour
...
+        // To let the punching animation finish 
+        if (!Punching && stateInfo.normalizedTime % 1 < 0.1f)
+        {
+            Animator.SetLayerWeight(PunchLayerIndex,0.0f);
+        }
+
+        if (Indestructibles.PlayerData.IsKnockedOut)
+        {
+            _stateMachine.StartWandering();
+        }
```
Comment (ClaudiaLapalme) [inline Assets/Scripts/NPCs/NPCManager.cs:67]: Do we even need to change the state of the NPCs? The game ends when the player is knocked out

File: Assets/Scripts/Player/PlayerController.cs [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+namespace Player
+{
+    public class PlayerController : MonoBehaviour
+    {
+        [SerializeField] private AudioClip[] gruntSounds;
+        [SerializeField] private GameManager gameManager;
+
+        [Header("Debug Variables")]
+        public bool invincible;
+        public bool onHitSoberUp;
```
Comment (ClaudiaLapalme) [inline Assets/Scripts/Player/PlayerController.cs]: Should probably be serialized?

File: Assets/Scripts/Player/PlayerController.cs [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+                    // Flicker effect
+                    InvokeRepeating(nameof(FlickerModel),0.0f,0.125f);
+                    Invoke(nameof(StopFlicker),2.0f);
+                    
+                    // Sober up
+                    if (onHitSoberUp)
+                    {
+                        Indestructibles.PlayerData.IntoxicationLevel -= 0.25f;
+                        ClearEffects();
+                    }
```
Comment (ClaudiaLapalme) [inline Assets/Scripts/Player/PlayerController.cs:153]: I think you forgot to remove this since we don't sober up on hit

File: Assets/Scripts/Structs/DebugControls.cs [hunk 11]
```diff
@@ -0,0 +1,18 @@
+﻿using UnityEngine;
+
+namespace Structs
+{
+    public struct DebugControls
+    {
+        public KeyCode DrinkABeer { get; set; }
+        public KeyCode ToggleInvincible { get; set; }
+        public KeyCode ToggleSoberOnHit { get; set; }
```
Comment (ClaudiaLapalme) [inline Assets/Scripts/Structs/DebugControls.cs:9]: Don't need this variable anymore

**Review by ClaudiaLapalme — CHANGES_REQUESTED** [review]
Great job! I love how everything in the architecture is a lot more organized. A few comments here and there, but otherwise looks good.

File: Assets/Scripts/NPCs/NPCManager.cs (hunk 3 above)
Comment (tedgag) [reply Assets/Scripts/NPCs/NPCManager.cs:114]: Disabled by default

File: Assets/Scripts/NPCs/NPCManager.cs (hunk 8 above)
Comment (tedgag) [reply Assets/Scripts/NPCs/NPCManager.cs:67]: This is just for better visuals, this way the NPC won't push on the dead player :)

File: Assets/Scripts/Player/PlayerController.cs (hunk 9 above)
Comment (tedgag) [reply Assets/Scripts/Player/PlayerController.cs]: Done

**Review by f-michaut — APPROVED** [review]

**Review by ClaudiaLapalme — APPROVED** [review]
